blog/_posts/2020-11-24-ranluxpp.md

```diff
@@ -0,0 +1,157 @@
+---
+title: "Implementing and tuning RANLUX++ for ROOT"
+layout: archive
+author: Jonas Hahnfeld
+---
+
+ROOT comes with support for different pseudorandom number generators (PRNGs).
+This post discusses the recent implementation of RANLUX++ and how I tuned its performance.
+
+The added `TRandomRanluxpp` joins the group of available RNGs in ROOT:
+
+ * [`TRandom`][TRandom] is a Linear Congruential Generator (LCG) based on the BSD `rand` generator.
+ * [`TRandom1`][TRandom1] implements the RANLUX generator described by Fred James in 1990.
+ * [`TRandom3`][TRandom3] is based on the Mersenne Twister generator.
+   By default, this implementation is used for `gRandom`.
+ * `TRandomMixMax` is an instantiation of the parametric [`MixMaxEngine`][MixMaxEngine].
+
+There are a few more variants that are linked [in the documentation][TRandom].
+
+[TRandom]: https://root.cern/doc/master/classTRandom.html
+[TRandom1]: https://root.cern/doc/master/classTRandom1.html
+[Trandom3]: https://root.cern/doc/master/classTRandom3.html
+[MixMaxEngine]: https://root.cern/doc/master/classROOT_1_1Math_1_1MixMaxEngine.html
+
+## Why so many PRNGs?
+
+There exist many properties to evaluate PRNGs, with the easiest being its period:
+it captures how many random numbers can be generated before the sequence "wraps around" and repeats.
+For obvious reasons, the period of a PRNG cannot be longer than its state.
+This is one of the problems of the oldest LCGs with a state of 32 bits or less.
+
+However, there are many more statistical properties that are at least equally important.
+A widely used collection of empirical tests is [TestU01][TestU01] by L'Ecuyer and Simard.
+It has to be noted that the tests can only prove the existence of defects.
+That is, passing all tests does not mean that a PRNG is free of errors.
+
+[TestU01]: http://simul.iro.umontreal.ca/testu01/tu01.html
+
+As with many other design issues, PRNGs need to make a certain trade-off:
+fast generators with a small state will usually fail many statistical tests.
+On the other hand, good generators are typically slower or use a much larger state.
+In fact, most generators will fail one test or another, including the widely used Mersenne Twister.
+Notable exceptions are RANLUX and MIXMAX that pass all currently known tests.
+However, they are rather slow (RANLUX) or require a large state (MIXMAX).
+
+## RANLUX++
+
+[RANLUX++][paper] is an LCG implementation derived from the ideas of RANLUX.
+It aims to provide better performance while maintaining the good properties of RANLUX.
+The generator uses a single 576 bit number as state.
+When invoked, it hands out the next unused bits that the generator remembers as an offset.
+Once all bits are exhausted, the LCG performs a multiplication and a modulo operation to advance the state.
+
+[paper]: https://arxiv.org/pdf/1705.03123.pdf
```

Comment:
I would add here a sentence mentioning that Ranlux++ allows to use a much higher luxury level (i.e. numbers which  needs to be skipped to obtain a chaotic behavior)  for free. For example the current implementation use a p=2048, much higher that the default of Ranlux (p=389).

Reply:
I intentionally avoided talking about luxury levels because you can only understand this after reading the referenced papers in depth. I wouldn't even claim that I fully get the relationship to the implementation of RANLUX++: the connection was obvious with RANLUX where skipping more states meant better quality, but in the LCG formulation it's just an exponentiation...

Reply:
I think some mention should be kept, because this is in my opinion a strength of this generator. Maybe we should not go into much detail and refer to the paper

Reply:
Sure, I added notes both here and in the conclusion that RANLUX++ uses more bits of randomness and has better properties "for free" (in terms of performance) that are understood in theory.